R/sample_grid.R

```diff
@@ -0,0 +1,109 @@
+##' Take a grid search produced by \code{\link{scan_beta_date}} and 
+##' sample \code{n_sample_pairs} from the parameter grid uses based
+##' on their probaility. For each parameter pair chosen, run particle
+##' filter with \code{num_particles} and sample 1 trajectory
+##' 
+##' @title Sample Grid Scan
+##' 
+##' @param scan_results Output of \code{\link{scan_beta_date}}.
+##' 
+##' @param n_sample_pairs Number of parameter pairs to be sampled. Integer.
+##'   Default = 10.
+##'   
+##' @param n_particles Number of particles. Positive Integer. Default = 100
+##'   
+##' @return List. First element is an array of trajectories, in which
+##'  first dimension is time, second is parameters chosen when sampling from 
+##'  the grid and the third dimension is alist of inputs.
+##' 
+##' @export
+##' @import furrr
+##' @importFrom utils tail
+sample_grid_scan <- function(scan_results, 
+                             n_sample_pairs = 10, 
+                             n_particles = 100) {
+  
+  # checks on args
+  assert_is(scan_results, "sircovid_scan")
+  assert_pos_int(n_sample_pairs)
+  assert_pos_int(n_particles)
+  
+  # grab the pobability matrix
+  prob <- scan_results$renorm_mat_LL
+  nr <- nrow(prob)
+  nc <- ncol(prob)
+  
+  # construct what the x and y dimensions look like
+  x_grid <- matrix(scan_results$x, nrow = nr, ncol = nc)
+  y_grid <- matrix(as.character(scan_results$y), nrow = nr, ncol = nc, byrow = TRUE)
+  
+  # draw which grid pairs are chosen
+  pairs <- sample(x =  length(prob), size = n_sample_pairs, 
+                  replace = TRUE, prob = prob)
+  
+  # what are related beta and dates
+  beta <- x_grid[pairs]
+  dates <- y_grid[pairs]
+  
+  # recreate parameters for re running
+  param_grid <- data.frame("beta" = beta, "start_date" = dates)
+  model_params <- scan_results$inputs$model_params
+  pars_obs <- scan_results$inputs$pars_obs
+  data <- scan_results$inputs$data
+  
+  
+  # Multi-core futures with furrr (parallel purrr)
+  
+  ## Particle filter outputs
```

Comment:
This is identical to the code in `grid_search.R`. Can you make this into its own function, which is then called in each of these two places

Reply:
Yeah we could do that actually as i just stole it. Will do that